src/rewrite/vct/rewrite/PrepareByValueClass.scala

```diff
@@ -0,0 +1,333 @@
+            if v.t.isInstanceOf[TNonNullPointer[Pre]] =>
+          val t = v.t.asInstanceOf[TNonNullPointer[Pre]]
+          if (t.element.isInstanceOf[TByValueClass[Pre]]) {
+            val newV: Ref[Post, HeapVariable[Post]] = succ(v)
+            val newP = dispatch(p)
+            Perm(HeapVariableLocation(newV), newP) &* Perm(
+              PointerLocation(DerefHeapVariable(newV)(dhv.blame))(pl.blame),
+              newP,
+            )
+          } else { node.rewriteDefault() }
+        case assign: PreAssignExpression[Pre] =>
+          val target = inAssignment.having(()) { dispatch(assign.target) }
+          if (assign.target.t.isInstanceOf[TByValueClass[Pre]]) {
+            copyContext.having(InAssignmentExpression(assign)) {
+              assign.rewrite(target = target)
+            }
+          } else {
+            // No need for copy semantics in this context
+            copyContext.having(NoCopy()) { assign.rewrite(target = target) }
+          }
+        case invocation: Invocation[Pre] =>
+          invocation.rewrite(args = invocation.args.map { a =>
+            if (a.t.isInstanceOf[TByValueClass[Pre]]) {
+              copyContext.having(InCall(invocation)) { dispatch(a) }
+            } else { copyContext.having(NoCopy()) { dispatch(a) } }
+          })
+        case dp @ DerefPointer(HeapLocal(Ref(v)))
+            if v.t.asPointer.get.element.isInstanceOf[TByValueClass[Pre]] =>
+          rewriteInCopyContext(
+            dp,
+            v.t.asPointer.get.element.asInstanceOf[TByValueClass[Pre]],
+          )
+        case dp @ DerefPointer(DerefHeapVariable(Ref(v)))
+            if v.t.asPointer.get.element.isInstanceOf[TByValueClass[Pre]] =>
+          rewriteInCopyContext(
+            dp,
+            v.t.asPointer.get.element.asInstanceOf[TByValueClass[Pre]],
+          )
+        case deref @ Deref(_, Ref(f)) if f.t.isInstanceOf[TByValueClass[Pre]] =>
+          if (copyContext.isEmpty) { deref.rewriteDefault() }
+          else {
+            // TODO: Improve blame message here
+            copyClassValue(
+              deref.rewriteDefault(),
+              f.t.asInstanceOf[TByValueClass[Pre]],
+              f => deref.blame,
+            )
+          }
+        case dp @ DerefPointer(Local(Ref(v)))
+            if v.t.asPointer.get.element.isInstanceOf[TByValueClass[Pre]] =>
+          // This can happen if the user specifies a local of type pointer to TByValueClass
+          rewriteInCopyContext(
+            dp,
+            v.t.asPointer.get.element.asInstanceOf[TByValueClass[Pre]],
+          )
+        case _ => node.rewriteDefault()
+      }
+  }
+
+  private def rewriteInCopyContext(
+      dp: DerefPointer[Pre],
+      t: TByValueClass[Pre],
+  ): Expr[Post] = {
+    if (copyContext.isEmpty) {
+      // If we are in other kinds of expressions like if statements
+      return dp.rewriteDefault()
+    }
+    val clazz = t.cls.decl.asInstanceOf[ByValueClass[Pre]]
```

Comment:
Maybe you can use "cls" here instead? Bit of a bikeshed maybe but I've always disliked the weird variations on class/this...